RavenDB-22457 Fixing shared referenced between transactions in next gen pager

ayende opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-RavenDB-22457

### Additional description

We accidently shared a `TreeRootHeader` for the root tree across all transactions (ang thus also held up transactions instances for too long)

That meant that changing of the root tree root page (whenever we split) would be _immediately_ visible to all transactions, leading to Bad Things.

This fixed the issue and ensure that we are using a struct, not a class, from now on. Thus, we don't have the option of repeating this issue.

ayende committed:
RavenDB-22457 - Adding a test showing how we share the mutable root tree

ayende committed:
RavenDB-22457 - Removing TreeMutableState because sharing its instance led to:
* Holding the transaction instance for too long
* Exposing the *uncommitted* state to read transactions

ayende committed:
RavenDB-22457 - Removing TreeMutableState from the tests

ayende committed:
RavenDb-22457
- Ensuring that we store a *copy* of the root tree state in the transaction header and the published record
- Reversing polarity of initializing the root tree

redknightlois committed:
RavenDB-22457 - Fixed benchmarking code.

redknightlois commented:
Check this one now: https://issues.hibernatingrhinos.com/issue/RavenDB-22859/Jalchr3.Streamingdocumentswillrespectthesortingorder because it fails consistently.

ayende committed:
RavenDB-22457
- We need to report the *total number of pages* for the pagers, because our metrics rely on it.
- Will need to think about how to properly do reports on that when we integrate with the Studio
